playbooks/robusta_playbooks/oom_killer.py

```diff
@@ -37,43 +55,211 @@ def oom_killer_enricher(event: NodeEvent):
+    limits: Optional[str] = None
+
+
+@dataclass
+class OomKill:
+    time: float
+    pod_name: str
+    container_name: str
+    image: str
+    memory_specs: MemorySpecs
+    reason: Optional[str]
+
+
+# This class is used only because the OomKillsExtractor should appear before KubernetesOomKillReasonInvestigator, but still know its interface.
+class OomKillReasonInvestigator(metaclass=abc.ABCMeta):
+    @abstractmethod
+    def get_reason(self, oom_kill: OomKill) -> str:
+        return ""
+
+
+class OomKillsExtractor:
+    def __init__(self, config: OomKillerEnricherParams, node: Node, oom_kill_reason_investigator: OomKillReasonInvestigator):
+        self.config = config
+        self.node = node
+        self.oom_kill_reason_investigator = oom_kill_reason_investigator
+
+        self.memory_transformer = K8sMemoryTransformer()
+
+    def extract_oom_kills(self) -> List[OomKill]:
+        results: PodList = Pod.listPodForAllNamespaces(
+            field_selector=f"spec.nodeName={self.node.metadata.name}"
+        ).obj
+
+        oom_kills: List[OomKill] = []
+        for pod in results.items:
+            pod_oom_kills = self.get_oom_kills_from_pod(pod)
+            for oom_kill in pod_oom_kills:
+                oom_kills.append(oom_kill)
+
+        return oom_kills
+
+    def get_oom_kills_from_pod(self, pod: Pod) -> List[OomKill]:
+        new_oom_kills_duration = timedelta(seconds=self.config.new_oom_kills_duration_in_sec)
+
+        containers_spec_by_name = {}
+        for c in pod.spec.containers:
+            containers_spec_by_name[c.name] = c
+
+        oom_kills: List[OomKill] = []
+        for c_status in pod.status.containerStatuses:
+            # Ignore pods that were not oom killed
+            state = self.get_oom_killed_state(c_status)
+            if state is None:
+                continue
+
+            # Ignore old oom kills
+            dt = parse_kubernetes_datetime_to_ms(state.terminated.finishedAt)
+            oom_kill_from = datetime.fromtimestamp(dt / 1000)
+            now = datetime.now()
+            if now - oom_kill_from > new_oom_kills_duration:
+                continue
+
+            # Report the current container as oom killed
+            resources = containers_spec_by_name[c_status.name].resources if c_status.name in containers_spec_by_name else None
+            memory_specs = self.get_memory_specs(resources)
+            oom_kill = OomKill(time=dt, pod_name=pod.metadata.name, container_name=c_status.name,
+                               image=c_status.image, memory_specs=memory_specs, reason=None)
+            oom_kill.reason = self.oom_kill_reason_investigator.get_reason(oom_kill)
+            oom_kills.append(oom_kill)
+
+        return oom_kills
+
+    def get_oom_killed_state(self, c_status: ContainerStatus) -> Optional[ContainerState]:
+        # Check if the container OOMKilled by inspecting the state field
+        if self.is_state_in_oom_status(c_status):
```

Comment:
Is this really possible? 
That the current status (not previous) is OOMKilled

Reply:
Yep. In fact, that was exactly why we failed to detect OOM kills exactly for that reason (See https://github.com/robusta-dev/robusta/pull/158).

To be honest, the distinction between lastState and state is not exactly clear to me, and I did notice that sometimes the lastState is in OOMKilled state, so the code is simply trying to check both.